synapse/rest/media/v1/filepath.py

```diff
@@ -80,7 +80,7 @@ def remote_media_thumbnail_rel(
         self, server_name, file_id, width, height, content_type, method
     ):
         top_level_type, sub_type = content_type.split("/")
-        file_name = "%i-%i-%s-%s" % (width, height, top_level_type, sub_type)
+        file_name = "%i-%i-%s-%s-%s" % (width, height, top_level_type, sub_type, method)
```

Comment:
won't this mean that all existing thumbnails get broken and orphaned?

Reply:
They will get orphaned. I'm not sure how to clean them up in this case. If they get broken, I don't know. At least my synapse regenerates them, but I also have dynamic thumbnailing enabled.

What would be a reasonable way to clean them up? Most thumbnails potentially are already broken with dynamic thumbnailing, as we can't really be sure, with what method they were generated and if they aren't truncated.

Reply:
They will get orphaned. I'm not sure how to clean them up in this case. If they get broken, I don't know. At least my synapse regenerates them, but I also have dynamic thumbnailing enabled.

What would be a reasonable way to clean them up? Most thumbnails potentially are already broken with dynamic thumbnailing, as we can't really be sure, with what method they were generated and if they aren't truncated.

Reply:
I agree that if you've got dynamic thumbnails enabled, then you have no real idea what is on the filesystem and you're best off removing them. (It looks like you're correct that the file will get recreated in this case). We should make sure not to orphan them, but that might be as simple as telling admins to delete the whole `remote_thumbnails` directory before upgrading, or give them a script to detect and delete files with the wrong name which they can run after the upgrade.

The problem is that most thumbnails are *not* broken when `dynamic_thumbnails` are disabled, so having them all orphaned (or even deleting them all so that they can be recreated is unacceptable).

A few options spring to mind:
1. record in the database somehow that these thumbnails are using the "legacy" filename
2. have the code fall back to the legacy filename if a file that the database says should exist is missing
3. try to figure out the likely method which was used for each thumbnail, and rename them accordingly.

I think option 2 is probably best?

Reply:
I think 2 is probably the easiest to implement and if marked correctly, it could probably be removed in the future, so I'll try to go for that. The rename in 3 would probably not be that much faster than thumbnailing images again as needed and would require a slow migration too. 1 would be doable, but would need another schema update in the future and I expect that table to be pretty big, so it may not be a good idea to add such baggage. :D

Reply:
Hm, I'm not so sure how to do any of this. I think I have an idea, how one could do method 3, if they didn't use dynamic thumbnailing (with dynamic you would need to wipe the cache anyway). You want to delete all your remote thumbnails if you had dynamic thumbnailing enabled, since you can't be sure, if your thumbnails are broken or not. If you didn't enable dynamic thumbnails, you can rename your remote thumbnails according to the methods in your homeserver.yaml, if they don't overlap. This takes a considerable disk overhead though and it may be easier to just delete them entirely and let them be recreated on demand, so that you don't need to rename a few million files.

Method 1 is pretty hard, if you want to handle all configurations correctly, since you can't register the "legacy" name for dynamic thumbnails without sometimes serving broken ones. It also now needs an additional db lookup to return the filename, which may lead to performance regressions.

Method 2 adds additional file system accesses for every cache miss and for legacy thumbnails, since you need to do 2 stats instead of 1.

So I probably prefer going with Method 3 and provide a script to rename the files or to just recommend to wipe the remote thumbnails to users. Otherwise I may need some help to implement 1 or 2.

I've already struggled quite a bit to understand how the background index update is supposed to work and I'm really not familiar enough with synapse/python to be confident to make such more complex changes.

Reply:
> Method 2 adds additional file system accesses for every cache miss and for legacy thumbnails, since you need to do 2 stats instead of 1.

I don't think this is correct. If it's a cache miss, the media will not be in the database, so we won't even bother looking at the filesystem. For legacy thumbnails, it's true that we'll end up with two stats, but I think that's a price we can pay for now.

If it's a real problem, we can write a separate change later to work through the database and rename any files with the wrong name (or to rename them as we find them). But for now I think we can keep things simpler by checking both filenames.